docs/rfcs/xxxx-backwards-compatibility.md

```diff
@@ -0,0 +1,111 @@
+
+## Goals 
+
+- Set expectations for what is covered by backwards compatibility. 
+- Set expectations for when upgrades to new major versions will be forced.
+
+## Proposal 
+
+Backwards compatibility means that a user can upgrade their version of Grafana Agent without needing to make any changes with the way they interact with Grafana Agent, provided that interaction is within scope of being covered by backwards compatibility.
+
+### Scope of backwards compatibility   
+
+The following will be protected by backwards compatibility between minor releases: 
+
+- **User configuration**, including configuration file syntax and the command-line interface. 
+
+- **Versioned network APIs**, if any versioned APIs are introduced prior to the 1.0 release.
+
+- **Telemetry data used in official dashboards**. This means that users will continue to be able to use the same set of dashboards we provide when upgrading minor releases.    
+
+  - Official dashboards are dashboards in the repository's `operations/` [directory](../../operations/).
+
+- **Externally importable Go packages**. If a user is importing our code as a dependency, they should be able to upgrade to a new minor release without having to make changes to their code.
+
+  This means that names, signatures, and meanings may not change; a function called Calculate may not be changed from returning the sum of two numbers to returning the difference of two numbers.
+
+  Protecting Go packages with backwards compatibility is expected to be a contentious proposal. However, it has positive benefits: 
+
+  - It makes us good open-source citizens, allowing people to use Grafana Agent as a library with the same upgrade confidence as people who use Grafana Agent as a binary.  
+
+  - It allows us to get in the habit of being deliberate around when something should be consumable downstream, and will help us strengthen our skills with designing good APIs in general. Ideally, this helps us get better at writing Go code _and_ writing Flow components.
+
+  - It will help train us to more easily recognize when any kind of change is breaking. 
+
+  The easiest way to commit to this without any effort is to put everything in an top-level internal folder, effectively saying that Grafana Agent has no code API that can be consumed downstream.
+
+- **The scope of backwards compatibility**. Backwards compatibility is only defined for major version 1; we reserve the right to change the definition of backwards compatibility between major versions. 
+
+If a breaking change is introduced in a minor change accidentally, and that breaking change is not covered by one of the [exceptions][] defined below, it is a bug. In these cases, a patch release should be introduced to undo the breaking change. 
+
+[exceptions]: #exceptions-to-backwards-compatibility
+
+### Exceptions to backwards compatibility 
```

Comment:
I'd like to add two new exceptions: 

1. Default values are not subject to backwards compatibility and may change between minor releases. 
2. Deprecating and making something a no-op is not a breaking change because the old configuration file is still accepted.

Reply:
> Default values are not subject to backwards compatibility and may change between minor releases.

I suppose the reasoning is that the previous default value would still be configurable, even though it is not a default.

It would be good if backwards compatibility gives an answer to the question "if I upgrade the Agent, will my old config continue to work in the same way?". If we change default values, the answer to that would be "no", so IMO it'd be better to not consider this backwards compatible.

Reply:
There is some nuance here on what backward compatibility means, IMO the document seems to indicate though maybe not spelled out that; backward compatibility means "the Agent will run but the behavior may change without changes". Or more simply "the inputs (config/api) won't change but the output of the Agent is not guaranteed".

Reply:
> Default values are not subject to backwards compatibility and may change between minor releases.

I think default values should be part of the backwards compatibility guarantee. It can completely break you if we e.g. change the default port of some component. When I depend on software and I see it's using semver and its a minor release, I expect it to not break anything and I am happy to quickly apply the update. If the software starts breaking with minor releases, I will lose trust and will not want to update often any more. 

IMO we should make default values part of the guarantee and be very careful about what we set up as default, because it will need to stay this way until the next breaking changes opportunity.

Reply:
> Deprecating and making something a no-op is not a breaking change ...

IMO making something a no-op is a breaking change, because e.g. my telemetry pipeline may stop working after a minor version update. I think that should not be the case - I'll quickly loose trust in minor updates if this is the case. 

Deprecation is not the same thing as making something a no-op though, we _can_ mark things as deprecated in minor releases - in docs (and maybe print a warning) and remove the feature only in the next major release. [See here](https://semver.org/#how-should-i-handle-deprecating-functionality). 

> not a breaking change because the old configuration file is still accepted

I don't think we should go with this definition of what is in scope of backwards compatibility guarantee. Imagine if k8s made a minor release where `Deployment` resource in your config file would become a no-op. It would be a breaking change. 

For me the TL;DR; should be that if I am using non-experimental, documented features, I should always be able to make a minor or a bugfix update without any extra work - just a version bump and nothing should break. And I don't need to read through changelog etc.

Reply:
> I think default values should be part of the backwards compatibility guarantee. It can completely break you if we e.g. change the default port of some component. 

The rationale behind the default values is to allow us to change things like the default value for the frequency at which the WAL gets truncated, but not things like port numbers. 

Maybe the language should be tweaked here a bit.

Reply:
> like the default value for the frequency at which the WAL gets truncated

I see - so this would be some knob for which we provide reasonable defaults, but allow customers to change as well if needed. I think for these we could consider one of the following options:
* Describe the default as `auto` and not really document what value it will be. If this changes, it'd be an undocumented change.
* Call out in docs that this configuration property may be tuned with minor releases. That should also be a property that only affects performance, but not the functionality. 
* Have a namespace/prefix for config properties that may change outside of the backwards compatibility guarantees. 

A general statement about defaults does sound like we give ourselves permission to break things like ports etc, so I think we should tweak this.

Reply:
I'm going to say that denoting that the "functional semantics" as being covered by backwards compatibility implies that non-functional semantics (i.e., performance tunable settings) are not covered. 

Marking as resolved.